Add Haus-Bus integration

Zero545 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Haus-Bus integration for devices of the German manufacturer [Haus-Bus.de](https://haus-bus.secure-stores.de/). This PR only integrates the Light based devices. Other types of devices will be added after this PR is accepted.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31840

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Zero545 committed:
Initial implemenation of hausbus light platform

Zero545 committed:
Moved is_light_channel to HausbusLight class

Zero545 committed:
Add test for config flow

Zero545 committed:
Add test for class HausbusDevice

Zero545 committed:
typo

Zero545 committed:
Simplify device test by using parametrize.
Add gateway tests

Zero545 committed:
return add channel future, so it can be awaited

Zero545 committed:
Did not work

Zero545 committed:
Return None as channel list for unknown device ids.
Separate initialization from status request in lights.

Zero545 committed:
Add tests for channel and gateway

Zero545 committed:
Move redundant code to helpers.py

Zero545 committed:
complete gateway test coverage

Zero545 committed:
Add first light tests

Zero545 committed:
Add more light tests. Fix for LED light type

Zero545 committed:
100 % test coverage

Zero545 committed:
Fix LED off command.
Add logger instance.

Zero545 committed:
Add pyhausbus reqirement, update strict typing and requirements

balloob commented on homeassistant/components/hausbus/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "homekit": {},
+  "iot_class": "local_push",
+  "loggers": ["pyhausbus"],
+  "requirements": ["pyhausbus==1.0.0"],
```
It looks like this is not a generic requirement but written specifically for Home Assistant (it has a `homeassistant` package). Why is this necessary and what is Home Assistant specific about it?

balloob replied:
I also don't see the source on GitHub linked from the PyPI package. Is there?

Zero545 replied:
The pyhausbaus package is maintained by the owner of [Haus-Bus.de](https://github.com/hausbus). The source code repository is located [here](https://github.com/hausbus/homeassistant). I am not sure, why it is not shown on the pypi page. I did add the link to the repo in the [pyproject.toml](https://github.com/hausbus/homeassistant/blob/main/pyproject.toml). 

There is nothing home assistant specific in that repository. It was named homeassistant, because the python library to access the Haus-Bus devices was created specifically for home assistant. But it is also possible to use it standalone. The python library offers an object to connect and communicate with the Haus-Bus devices via UDP.

Zero545 replied:
pypi page has been updated with a 1.0.1 release of pyhausbus

zweckj commented on homeassistant/components/hausbus/channel.py:
```diff
@@ -0,0 +1,48 @@
+"""Representation of a Haus-Bus channel."""
```
This file is usually called `entity.py`

Zero545 replied:
I will rename it.

zweckj commented on homeassistant/components/hausbus/channel.py:
```diff
@@ -0,0 +1,48 @@
+        self._instance_id = instance_id
+        self._device = device
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique identifier for this channel."""
+        return f"{self._device.device_id}-{self._type}{self._instance_id}"
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Return a device description for device registry."""
+        return self._device.device_info
+
+    @property
+    def translation_key(self) -> str | None:
+        """Return the translation key to translate the entity's states."""
+        return self._type
+
+    @property
+    def name(self) -> str | None:
+        """Return the channel name."""
+        return f"{super().name} {self._instance_id}"
```
please use the attributes `attr_...` instead

zweckj commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    _LOGGER.debug("start searching devices")
+
+    # search devices after adding all callbacks to the gateway object
+    gateway.home_server.searchDevices()
```
can't this throw exceptions? Can we do this before we setup the entries, so we abort if anything goes wrong?

Zero545 replied:
This does essentially only send a UDP broadcast into the network to trigger the device discovery. The actual devices will be populated by events that are received in response to this.

zweckj commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,43 @@
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for hausbus."""
+
+    VERSION = 1
```
```suggestion
```
is the default

zweckj commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,43 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
```
```suggestion
```
it seems like you're not doing anything here, so no errors possible. But I would argue that I'd make sense to check if we even have devices here first.

Zero545 replied:
Devices are added asynchronously. We are just setting things up and sending a UDP broadcast to the network for discovery afterwards. So at this step we do not have any devices.

zweckj commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,158 @@
+"""Representation of a Haus-Bus device."""
```
this belong to the library.

zweckj replied:
I'm not saying the classes should just be dropped in the lib as-is. It just seems to me, like there's lot of things in those 3 files (like the binary mapping in this file) that are not really HA specific. Bases for all entities are usually placed in a file called `entity.py`

Zero545 replied:
I guess you are right about the mapping. I will ask the library owner to move it there.

The device class is not the base for all entities. These are devices that can have multiple entities. It is not a one to one relation for devices to entities.

Zero545 replied:
Moved the mapping to the pyhausbus library

zweckj commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,158 @@
+                match type_id:
+                    case 0x18:
+                        self.model_id = "6-fach Taster"
+                    case 0x19:
+                        self.model_id = "4-fach Taster"
+                    case 0x1A:
+                        self.model_id = "2-fach Taster"
+                    case 0x1B:
+                        self.model_id = "1-fach Taster"
+                    case 0x1C:
+                        self.model_id = "6-fach Taster Gira"
+                    case 0x20:
+                        self.model_id = "32-fach IO"
+                    case 0x0C:
+                        self.model_id = "16-fach Relais"
+                    case 0x08:
+                        self.model_id = "8-fach Relais"
+                    case 0x10:
+                        self.model_id = "24-fach UP-IO"
+                    case 0x28:
+                        self.model_id = "8-fach Dimmer"
+                    case 0x29:
+                        self.model_id = "8-fach Dimmer"
+                    case 0x30:
+                        self.model_id = "2-fach RGB Dimmer"
+                    case 0x00:
+                        self.model_id = "4-fach 0-10V Dimmer"
+                    case 0x01:
+                        self.model_id = "4-fach 1-10V Dimmer"
+                    case _:
+                        self.model_id = "Controller"
+            case EFirmwareId.SD485:
+                match type_id:
+                    case 0x28:
+                        self.model_id = "24-fach UP-IO"
+                    case 0x1E:
+                        self.model_id = "6-fach Taster"
+                    case 0x2E:
+                        self.model_id = "6-fach Taster"
+                    case 0x2F:
```
duplicate code in here. Keep it in english please.

Zero545 replied:
These are the official product names, but I will try to translate.

zweckj replied:
ideally, you'd just set translation_keys and add the (english) names to `strings.json` so they can get translated automatically https://developers.home-assistant.io/docs/internationalization/core#devices

zweckj commented on homeassistant/components/hausbus/event_handler.py:
```diff
@@ -0,0 +1,19 @@
+"""Event handler abstract class."""
```
probably belongs to the library

Zero545 replied:
This is just an abstract class/function call that we can use to set up an entity at runtime, when we receive the event from the library that a new entity was detected.

zweckj commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,156 @@
+"""Representation of a Haus-Bus gateway."""
```
this also looks like something that belongs in the library

Zero545 replied:
This is the part that actually translates to library events to home assistant specific stuff, so I'd say this should not be in the pyhausbus library.

zweckj commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,274 @@
+class HausbusLight(HausbusChannel, LightEntity):
+    """Representation of a Haus-Bus light."""
+
+    TYPE = DOMAIN
```
```suggestion
```

zweckj commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,274 @@
+    gateway.register_platform_add_channel_callback(async_add_light, DOMAIN)
+
+
+class HausbusLight(HausbusChannel, LightEntity):
```
if you split this into a `HausBusLight` and from that inherit then `HausBusDimmerLight`, `HausBusLEDLight`, `HausBusRGBLight` classes you can avoid all those `if`s, because you have a specific class per type

Zero545 replied:
This was easier than expected, thanks.

zweckj commented on homeassistant/components/hausbus/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "loggers": ["pyhausbus"],
+  "requirements": ["pyhausbus==1.0.0"],
+  "single_config_entry": true,
+  "ssdp": [],
+  "zeroconf": []
```
remove any empty keys

zweckj reviewed (changes requested):
Thanks for your contribution! 👍 Some feedback for you

LennP commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,158 @@
+            via_device=(DOMAIN, self.bridge_id),
+        )
+
+    def set_type(self, type_id: int):
+        """Set device name and model_id according to device type."""
+        match self.firmware_id:
+            case EFirmwareId.ESP32:
+                match type_id:
+                    case 0x65:
+                        self.model_id = "LAN-RS485 Brückenmodul"
```
Instead of using nested match cases I would use a dictionary like this:

```python
model_ids = {
  EFirmwareID.ESP32: {
    0.65: "LAN-RS485 Brückenmodul",
    ...
  },
  ...
}
``` 

And then use `get` to get the default value of "Controller" in this case:

```python
firmware_model_id = model_ids.get(self.firmware_id, "Controller")
self.model_id = firmware_id if type(firmware_model_id) is str else firmware_model_id.get(type_id, "Controller")
```

LennP replied:
Or indeed as zweckj said try to provide translations. You will probably still need a dictionary for that though

LennP commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,156 @@
+
+    def add_light_channel(self, instance: ABusFeature, object_id: ObjectId):
+        """Add a new Haus-Bus Light Channel to this gateways channel list."""
+        device = self.get_device(object_id)
+        if device is not None:
+            light = HausbusLight(
+                object_id.getInstanceId(),
+                device,
+                instance,
+            )
+            channel_list = self.get_channel_list(object_id)
+            if channel_list is not None:
+                channel_list[self.get_channel_id(object_id)] = light
+                asyncio.run_coroutine_threadsafe(
+                    self._new_channel_listeners[LIGHT_DOMAIN](light), self.hass.loop
+                ).result()
+                light.get_hardware_status()
```
You can reduce nesting by returning if device is None, same for channel_list

LennP commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,274 @@
+    @callback
+    async def async_add_light(channel: HausbusChannel) -> None:
+        """Add light from Haus-Bus."""
+        entities: list[HausbusLight] = []
+        if isinstance(channel, HausbusLight):
+            entities.append(channel)
+        async_add_entities(entities)
```
```suggestion
        if isinstance(channel, HausbusLight):
            async_add_entities([channel])
```

Zero545 committed:
Remove empty keys from manifest.json
Remove unused TYPE definition in light.py
Remove ConfigFlow VERSION
Change async_step_user return type to ConfigFlowResult

Zero545 committed:
Simplify async_add_light callback

emontnemery commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,46 @@
+from .const import DOMAIN
+from .gateway import HausbusGateway
+
+# Only support light platform on the initial version. Additional platforms to follow
```
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    hass.data.setdefault(DOMAIN, {})
+
+    gateway = HausbusGateway(hass, entry)
+    hass.data[DOMAIN][entry.entry_id] = gateway
```
Instead of this, implement https://developers.home-assistant.io/blog/2024/04/30/store-runtime-data-inside-config-entry

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,42 @@
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            return self.async_create_entry(title="Haus-Bus", data=user_input)
```
We should try to discover haus-bus devices and only create the config entry if devices were found.
You can use `self.async_show_progress` to inform the user we're waiting for devices to be discovered.

Zero545 replied:
I did add a simple discovery in the config_flow. Now it will only add the hausbus config entry, if at least one device has been discovered.

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,42 @@
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
Remove these if they are not needed

emontnemery reviewed (changes requested):
There are a lot of unresolved comments from previous reviews, please fix them.
I also added a few additional comments.

Zero545 committed:
Fix config entry setup for tests

CTSchorsch committed:
fix some comments from pull request

CTSchorsch committed:
fix some more comments:
change runtime data handling
change case structures to dictonary

CTSchorsch committed:
save

CTSchorsch committed:
fix tests

Zero545 committed:
Moved device type resolution to pyhausbus library. Updated library required version.

Zero545 committed:
Split HausbusLight into 3 sub-classes HausbusDimmerLight, HausbusRGBDimmerLight, HausbusLedLight

Zero545 committed:
Extend config_flow to only add a haus-bus entry, if at least one haus-bus device has been detected.

Zero545 committed:
Add tests for new config_flow

emontnemery commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,43 @@
+    _LOGGER.debug("start searching devices")
+
+    # search devices after adding all callbacks to the gateway object
+    gateway.home_server.searchDevices()
```
If this sends UDP packages, this should be run in an executor:
```suggestion
    hass.async_add_executor_job(gateway.home_server.searchDevices)
```

Also, shouldn't this happen periodically, every minute or something instead of once? What happens if the device is disconnected when Home Assistant starts?

Zero545 replied:
Yes, if the Hausbus LAN-Gateway is not connected during Home Assistant startup, all Hausbus devices will be shown as unavailable until you reload the integration. I am unsure if the overhead of a periodic job is worth to handle this corner case, or if it would be enough to add this detail to the documentation. Is there an easy way to schedule a periodic job in home assistant?

Other Hausbus devices on the bus can be detected at runtime, even if they are connected or boot later.

emontnemery replied:
It's not OK to require Home Assistant to be restarted.

You write "the Hausbus LAN-Gateway", does it mean a user will always have exactly one gateway, or could a user also have no gateway or more than one gateway?

emontnemery replied:
> you can reload the integration from the integrations page

It should not be necessary for the user to perform such manual steps (restart Home Assistant or reload the integration) for a basic and common issue where the device is not connected to the network when Home Assistant starts.

> They need to have one gateway connected to their LAN network. There are multiple devices that can serve as a gateway.

What happens if there are more than one gateway on the LAN network, will Home Assistant connect to one of them based on the UDP discovery?
What happens if the connection to the gateway is lost?

Zero545 replied:
Since the communication is UDP there is no long lasting connection. It will just send and receive command and events from the network. If the user connects multiple gateways for some reason, the events would be received multiple times, but this is no problem.
If the LAN gateway got disconnected the commands will not reach the destination and there will be no state update events. So the devices are frozen in the last state from Home Assistant perspective.

Zero545 replied:
I also implemented a recurring background job for device discovery. I hope I did it in the correct way.

emontnemery replied:
> Since the communication is UDP there is no long lasting connection. It will just send and receive command and events from the network.

Yeah, that's clear 👍 

>  If the user connects multiple gateways for some reason, the events would be received multiple times, but this is no problem.

OK, so if there are more than one gateway on the LAN, the single hausbus config entry take care of them all?

> If the LAN gateway got disconnected the commands will not reach the destination and there will be no state update events. So the devices are frozen in the last state from Home Assistant perspective.

Right, that makes sense. But once they're back online, everything will again work from the user's perspective without manual intervention?

> I also implemented a recurring background job for device discovery. I hope I did it in the correct way.

It looks good 👍

Zero545 replied:
> OK, so if there are more than one gateway on the LAN, the single hausbus config entry take care of them all?
Yes.

>  But once they're back online, everything will again work from the user's perspective without manual intervention?
Correct.

emontnemery commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,43 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: HausbusConfigEntry) -> bool:
+    """Unload a config entry."""
+    # gateway = cast(HausbusGateway, hass.data[DOMAIN][entry.entry_id])
```
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/config_entry.py:
This is not Java, we don't need to create separate modules for every class. I'd suggest to move this to `__init__.py` instead.

Zero545 replied:
I am running into circular import problems, if I do this.

emontnemery replied:
Right, this is a very big weakness of Python.

There are two ways around it:
1. If you only need the type for type annotations, import behind a `TYPE_CHECKING` guard:
```py
if TYPE_CHECKING:
    import xxx
```
2. Import locally, in a class or function - the linter will complain, but it's fine to silence the error if you're importing to avoid circular imports.

emontnemery replied:
Please don't mark comments as resolved without a) adressing the comment or b) explaining why you don't agree.

In this case, the solution is very simple:
1. Move HausbusConfigEntry and HausbusConfig to `__init__.py`
2. In `light.py` change like this:
```py
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import ATTR_ON_STATE
from .device import HausbusDevice
from .entity import HausbusEntity

if TYPE_CHECKING:
    from . import HausbusConfigEntry
```

Zero545 replied:
Thanks. This did not work at first for me. But after adding the suggested `from __future__ import annotations` this is working flawlessly.

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+        if not self._search_task.done():
+            return self.async_show_progress(
+                step_id="wait_for_device",
+                progress_action="wait_for_device",
```
There are no translations for the step_id + action, please add that, otherwise I think the user only sees a spinner with no explanation while searching.

Zero545 replied:
Are there any resources regarding how to correctly use translations? I found it hard to find the correct keys in the strings.json.

emontnemery replied:
There's some explanations here: https://developers.home-assistant.io/docs/data_entry_flow_index#labels--descriptions

When you test locally, you need to make sure `python3 -m script.translations develop --all` is executed before `hass`.
Also, you may need to force-refresh your browser for it to re-fetch translations.

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+        if user_input is not None:
+            return await self.async_step_wait_for_device()
+
+        self._set_confirm_only()
```
Why do we set this here?

Zero545 replied:
I copied this from an other integration, will check if this is needed at all.

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+    async def _async_wait_for_device(self) -> None:
+        """Start searching for devices and wait until at least one device was found or timeout is reached."""
+        self.home_server.searchDevices()
```
I guess this sends UDP packages, so should be run in an executor thread, not an event loop?

Zero545 replied:
The event loop is handled in the pyhausbus library.

Zero545 replied:
Can you tell me how to access the hass executor thread from here? In __init__.py I had access to the hass object to add the executor job.

emontnemery replied:
The hass object is available as `self.hass`

emontnemery commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+        """Initialize the config flow."""
+        self._found_device = False
+        self._search_task: asyncio.Task | None = None
+        self.home_server = HomeServer()
```
Do we need to take any care to shut down the `HomeServer` object when ending the flow?

Zero545 replied:
This is a good point. For completeness we should remove the event listeners.

Zero545 replied:
I tried to add a method to remove the bus event listeners and call it from the search_complete step, but how do I call it in case the flow is aborted?

emontnemery replied:
You need to override this method https://github.com/home-assistant/core/blob/2462dfbc473232aa40ebcaf5861e0bedec6c256e/homeassistant/data_entry_flow.py#L881

emontnemery commented on homeassistant/components/hausbus/const.py:
```diff
@@ -0,0 +1,7 @@
+LOGGER = logging.getLogger(__package__)
+
+DOMAIN = "hausbus"
+ATTR_ON_STATE = "on_state"
```
Only used in `light.py`, please move it there unless you will use it from other modules in a follow-up

Zero545 replied:
The idea was to also use it for other platforms like switch.

emontnemery replied:
Let's not worry about the future before we are there

emontnemery commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,58 @@
+        self.bridge_id = bridge_id
+        self.firmware_id = firmware_id
+
+    @property
+    def device_id(self) -> str | None:
+        """Return a serial number for this device."""
+        return self._device_id
```
Why do we type a function returning a `str` as `str | None`?
Also, the getter seems useless, I'd suggest to remove it.

emontnemery replied:
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,58 @@
+        firmware_id: EFirmwareId,
+    ) -> None:
+        """Set up Haus-Bus device."""
+        self._device_id = device_id
```
```suggestion
        self.device_id = device_id
```

emontnemery commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,58 @@
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Return a device description for device registry."""
+        if self.device_id is None:
+            return None
+
```
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,58 @@
+        return self._device_id
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
```
```suggestion
    def device_info(self) -> DeviceInfo:
```

emontnemery commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,35 @@
+    @callback
+    def async_update_callback(self, **kwargs: Any) -> None:
+        """State push update."""
+        raise NotImplementedError
```
As an alternative, you could make this an `@abstractmethod` if all subclasses will implement `async_update_callback`

emontnemery commented on homeassistant/components/hausbus/event_handler.py:
```diff
@@ -0,0 +1,20 @@
+from .entity import HausbusEntity
+
+
+class IEventHandler(ABC):
```
This is not Java, move this to `gateway.py`

Zero545 replied:
This will also create circular dependencies. I moved it outside of gateway.py to define the interface between the gateway and the platform entities.

emontnemery replied:
How can there be a circular dependency when the class is only referenced in `gateway.py`?

Zero545 replied:
You are right. After the gateway instance moved to the HausbusConfig entry the circular dependency was already solved. This also means that the IEventHandler class is not even necessary and I will completely remove it.

emontnemery commented on homeassistant/components/hausbus/event_handler.py:
```diff
@@ -0,0 +1,20 @@
+        platform: str,
+    ) -> None:
+        """Register add device callbacks."""
+        raise NotImplementedError
```
It's not possible to instantiate a class which has unimplemented abstract methods, so this can be removed
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/event_handler.py:
```diff
@@ -0,0 +1,20 @@
+from .entity import HausbusEntity
+
+
+class IEventHandler(ABC):
```
Prefixing interface classes with `I` looks like a Java thing, we don't do that in this project, I suggest to change to:
```suggestion
class EventHandler(ABC):
```

OK to keep if you want to be consistent with the pyhausbus library though

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+
+from pyhausbus.ABusFeature import ABusFeature
+from pyhausbus.BusDataMessage import BusDataMessage
+from pyhausbus.de.hausbus.homeassistant.proxy.Controller import Controller
+from pyhausbus.de.hausbus.homeassistant.proxy.controller.data.Configuration import (
+    Configuration,
+)
+from pyhausbus.de.hausbus.homeassistant.proxy.controller.data.ModuleId import ModuleId
+from pyhausbus.de.hausbus.homeassistant.proxy.controller.data.RemoteObjects import (
+    RemoteObjects,
+)
```
I've never seen this Java-style package naming in a Python library 🤷‍♂️

Zero545 replied:
The creator of the library is a Java-developer. I think he generated most of the python library from his existing code base.

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+        """Initialize the system."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.bridge_id = "1"
```
Why is this hard coded like this? If bridge ID is always 1, maybe just remove it from the class?

Zero545 replied:
I thought I could make use of the bridge_id, but it turned out obsolete.

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+        ] = {}
+
+    def add_device(self, device_id: str, module: ModuleId) -> None:
+        """Add a new Haus-Bus Device to this gateways devices list."""
```
```suggestion
        """Add a new Haus-Bus Device to this gateway's device list."""
```

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+
+    def add_device(self, device_id: str, module: ModuleId) -> None:
+        """Add a new Haus-Bus Device to this gateways devices list."""
+        device = HausbusDevice(
+            self.bridge_id,
+            device_id,
+            module.getFirmwareId().getTemplateId()
+            + " "
+            + str(module.getMajorRelease())
+            + "."
+            + str(module.getMinorRelease()),
+            module.getName(),
+            module.getFirmwareId(),
+        )
+        if device_id not in self.devices:
+            self.devices[device_id] = device
```
Don't first create an object only to then check if it's needed

emontnemery replied:
Now `if device_id not in self.devices:` is checked twice, please combine it

Zero545 replied:
I don't think `if device_id not in self.devices:` is performed twice. There is an other check for the channels dict `if device_id not in self.channels:`, though. Do you mean I should combine the 2 dicts "devices" and "channels"? I would rather keep those separate, even if the entries for a specific device will normally always be created at the same time.

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+
+    def get_device(self, object_id: ObjectId) -> HausbusDevice | None:
+        """Get the device referenced by ObjectId from the devices list."""
+        return self.devices.get(str(object_id.getDeviceId()), None)
```
`dict.get` returns `None` if the key is missing, there's no need to pass `None`. Please change that here and in the other cases
```suggestion
        return self.devices.get(str(object_id.getDeviceId()))
```

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+        self, object_id: ObjectId
+    ) -> dict[tuple[str, str], HausbusEntity] | None:
+        """Get the channel list of a device referenced by ObjectId."""
+        return self.channels.get(str(object_id.getDeviceId()), None)
```
```suggestion
        return self.channels.get(str(object_id.getDeviceId()))
```

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+        return None
+
+    def add_light_channel(self, instance: ABusFeature, object_id: ObjectId) -> None:
+        """Add a new Haus-Bus Light Channel to this gateways channel list."""
```
```suggestion
        """Add a new Haus-Bus Light Channel to this gateway's channel list."""
```

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+                device,
+                instance,
+            )
+        return None
```
Should this raise instead if the type of light is unsupported?

Zero545 replied:
Are there any benefits? I think it would be handled the same way. The exception would also just be ignored.

emontnemery replied:
No, there's no special benefit.

Zero545 replied:
If there is no benefit, I would like to keep it as is.

emontnemery commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,185 @@
+        object_id = ObjectId(busDataMessage.getSenderObjectId())
+        data = busDataMessage.getData()
+
+        if object_id.getDeviceId() == 9998:
```
Please don't use magic numbers, add a `Final` constant with a clear name instead

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Haus-Bus lights from a config entry."""
+    # gateway = cast(IEventHandler, hass.data[HAUSBUSDOMAIN][config_entry.entry_id])
```
```suggestion
```

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+
+    def get_hardware_status(self) -> None:
+        """Request status of a light channel from hardware."""
+        raise NotImplementedError
```
should this be an `abstractmethod` instead?

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+        if isinstance(data, (DimmerEvOff, ledEvOff, rgbDimmerEvOff)):
+            channel.light_turn_off()
+
+    def turn_on(self, **kwargs: Any) -> None:
+        """Turn on action."""
+        raise NotImplementedError
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn off action."""
+        raise NotImplementedError
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on light."""
+        self.turn_on(**kwargs)
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn off light."""
+        self.turn_off()
```
We don't need to copy these from the base class

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+
+    def light_turn_off(self) -> None:
+        """Turn off a light channel."""
+        params = {
+            ATTR_ON_STATE: False,
+        }
```
```suggestion
        params = {ATTR_ON_STATE: False}
```

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+        self.async_update_callback(**params)
+
+    @staticmethod
+    def handle_light_event(data: Any, channel: HausbusEntity) -> None:
```
Why is this a static method instead of an instance method?

Zero545 replied:
There is not supposed to be any HausbusLight instance and the method is independent from any instance.

emontnemery replied:
I don't understand this explanation.

> There is not supposed to be any HausbusLight instance

Sure, there are no instances of the abstract base class, but there are instances of the concrete classes `HausbusDimmerLight`, `HausbusRGBDimmerLight` and `HausbusLedLight`.

> the method is independent from any instance

This is clearly not the case since we pass in a reference to a `HausbusEntity` which we on the very first line check is a `HausbusLight`?

If this was instead an instance method, we would call it like this from `HausbusGateway`:
```py
        channel = self.get_channel(object_id)
        if channel is not None:
            cast(HausbusLight, channel).handle_light_event(data, channel)
```

emontnemery commented on homeassistant/components/hausbus/light.py:
```diff
@@ -0,0 +1,306 @@
+        """Set up light."""
+        super().__init__(instance_id, device, channel)
+
+        self._channel = cast(Dimmer, channel)
```
Instead of casting here, could we type the `__init__` parameter as `Dimmer`?

Zero545 replied:
I did not think that this is possible, but everything still works as expected :+1:

emontnemery reviewed (changes requested):
Some comments. I've not reviewed the tests.

Zero545 committed:
Use executor job to send UDP request for device search.

Zero545 committed:
Update config flow text and remove event listeners after finishing the config_flow.

Zero545 committed:
Simplified HausbusDevice id and info

Zero545 committed:
Rename IEventHandler and remove "NotImplementedError"

Zero545 committed:
Removed bridge_id from Gateway and HausbusDevice

Zero545 committed:
Use constant for the Hausbus homeserver device id

Zero545 committed:
Apply simplifications to the light classes.

emontnemery commented:
Please fix the CI errors.
Also, if you want the PR to be reviewed again, please click the "Ready for review" button 👍

Zero545 commented:
I want to keep this PR open. I didn't have time lately to work on the requested changes.

Zero545 committed:
Do not set the on state before receivig the updated state from the Light device.

Zero545 committed:
Silence the import-outside-toplevel error, needed to avoid circular import.

Zero545 committed:
Perform device discovery periodically every minute.

Zero545 committed:
Set light by percent instead of value in 0..255.

emontnemery commented:
@Zero545 Please fix the lint errors
Also, feel free to contact me on discord, I'm @emontnemery there, if you need some help.

emontnemery commented:
Please add `from __future__ import annotations` to all modules, we use [PEP 563](https://peps.python.org/pep-0563/) extensively in this project as it greatly simplifies the use of forward references of types.

emontnemery reviewed (changes requested):
There are unresolved comments from previous reviews, please check them.

Also, there's some issue with the tests; they need a long time to complete, I think the problem is `test_form_user_timeout`. Please change that test to patch the timeout to 0, we don't need to wait 10s in the test for the timeout.

Zero545 committed:
Fix ruff-format and pylint errors

Zero545 committed:
Add "from __future__ import annotations" to all modules and move HausbusConfig to __init__.py

Zero545 committed:
Move removal of bus event listeners to asnyc_remove.

Zero545 committed:
Remove EventHandler class as it is obsolete.

Zero545 committed:
Apply code simplifications

Zero545 committed:
Patch timeout time for "test_form_user_timeout" to 0

Zero545 committed:
Change handle_light_event to an instance method and split event handling into the respective subclasses.

joostlek reviewed (changes requested):
New integration are required to have a `quality_scale.yaml` file and reach at least bronze. So I am going to put this back into draft, let's pick it up once it has been added

frenck committed:
Merge branch 'dev' into hausbus_lights

Zero545 commented:
By the way, I am currently working on the quality scale list.
